File: src/dp/knapsack/km46.cc
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int main() {
  int itemNum, totalVolume;
  cin >> itemNum >> totalVolume;

  auto itemVolumes = vector<int>(itemNum, 0);
  auto itemValues = vector<int>(itemNum, 0);

  for (auto i = 0; i < itemNum; i++) {
    cin >> itemVolumes[i];
  }

  for (auto i = 0; i < itemNum; i++) {
    cin >> itemValues[i];
  }

  auto dp = vector<vector<int>>(itemNum + 1, vector<int>(totalVolume + 1, 0));

  for (auto i = 1; i <= itemNum; i++) {
    for (auto j = 1; j <= totalVolume; j++) {
      if (j < itemVolumes[i - 1]) {
        dp[i][j] = dp[i - 1][j];
        continue;
      }

      dp[i][j] = max(dp[i - 1][j],
                     dp[i - 1][j - itemVolumes[i - 1]] + itemValues[i - 1]);
    }
  }

  cout << dp[itemNum][totalVolume] << endl;
}

File: src/dp/knapsack/km46.ts
import { createInterface } from "readline";

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lineno = 0;

let itemNum = 0;
let totalVolume = 0;
let itemVolumes: number[];
let itemValues: number[];

rl.on("line", (line) => {
  lineno++;

  if (lineno === 1) {
    [itemNum, totalVolume] = line
      .trim()
      .split(" ")
      .map((item) => Number.parseInt(item));
    return;
  }

  if (lineno === 2) {
    itemVolumes = line
      .trim()
      .split(" ")
      .map((item) => Number.parseInt(item));
    return;
  }

  itemValues = line
    .trim()
    .split(" ")
    .map((item) => Number.parseInt(item));
  console.log(maxTotalValue(itemNum, totalVolume, itemVolumes, itemValues));
  rl.close();
});

function maxTotalValue(
  itemNum: number,
  totalWeight: number,
  itemWeights: number[],
  itemValues: number[],
): number {
  // 初始化 dp 数组
  const dp = Array.from({ length: itemNum + 1 }, () =>
    new Array(totalWeight + 1).fill(0),
  );

  // console.log(dp)
  for (let i = 1; i <= itemNum; i++) {
    for (let j = 1; j <= totalWeight; j++) {
      if (j < itemWeights[i - 1]) {
        dp[i][j] = dp[i - 1][j];
        continue;
      }
      dp[i][j] = // 前 i 个物品, 放入容量为 j 的背包的最大收益
        Math.max(
          // 不放物品 i
          // 则 dp[i][j] = 前 i-1 个物品, 放入容量为 j 的背包的最大收益
          dp[i - 1][j],
          // 放物品 i
          // 则 dp[i][j] = 前 i-1 个物品, 放入容量为 j-w[i-1] 的背包的最大收益,
          // 再加上 物品 i-1 的收益 v[i-1]
          dp[i - 1][j - itemWeights[i - 1]] + itemValues[i - 1],
        );
    }
  }
  return dp[itemNum][totalWeight];
}
